partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt

```diff
@@ -62,13 +63,28 @@ internal class ExprCallDynamic(
             return fn.invoke(args)
         }
 
-        internal fun matches(args: Array<PartiQLValue>): Boolean {
-            for (i in args.indices) {
-                if (types[i] == PartiQLValueType.ANY) {
-                    return true
-                }
-                if (args[i].type != types[i]) {
-                    return false
+        internal fun matches(inputs: Array<PartiQLValue>): Boolean {
+            for (i in inputs.indices) {
+                val inputType = inputs[i].type
+                val parameterType = signatureParameters[i]
+                val c = coercions[i]
+                when (c) {
+                    // coercion might be null if one of the following is true
+                    // Function parameter is ANY,
+                    // Input type is null
+                    // input type is the same as function parameter
+                    null -> {
+                        if (!(inputType == parameterType || inputType == PartiQLValueType.NULL || parameterType == PartiQLValueType.ANY)) {
+                            return false
+                        }
+                    }
+                    else -> {
+                        // checking the input type is expected by the coercion
+                        if (inputType != c.input) return false
+                        // checking the result is expected by the function signature
+                        // this should branch should never be reached, but leave it here for clarity
+                        if (c.target != parameterType) return false
```

Comment:
```suggestion
                        if (c.target != parameterType) error("Internal Error: Cast Target does not match Function Parameter")
```

Reply:
Let's keep the branch here for now, as we are still working towards finalize the type Semantic and cast semantic. 

Let me know if this is sufficient for you.

Reply:
Yeah, that works!